app/logic/Contacts/ActiveSync/ActiveSyncAddressbook.ts

```diff
@@ -1,16 +1,129 @@
 import { Addressbook } from "../Addressbook";
+import { SQLAddressbook } from '../SQL/SQLAddressbook';
+import { SQLPerson } from '../SQL/SQLPerson';
 import { ActiveSyncPerson } from "./ActiveSyncPerson";
-import { ActiveSyncGroup } from "./ActiveSyncGroup";
-import type { ActiveSyncAccount } from "../../Mail/ActiveSync/ActiveSyncAccount";
+import { EASError, type ActiveSyncAccount, type ActiveSyncPingable } from "../../Mail/ActiveSync/ActiveSyncAccount";
+import { kMaxCount, ensureArray } from "../../Mail/ActiveSync/ActiveSyncFolder";
+import { NotSupported } from "../../util/util";
+import type { ArrayColl } from "svelte-collections";
 
-export class ActiveSyncAddressbook extends Addressbook {
+export class ActiveSyncAddressbook extends Addressbook implements ActiveSyncPingable {
   readonly protocol: string = "addressbook-activesync";
+  readonly persons: ArrayColl<ActiveSyncPerson>;
   account: ActiveSyncAccount;
+  syncKeyBusy: Promise<any> | null;
+  readonly folderClass = "Contacts";
+
+  get serverID() {
+    return new URL(this.url).searchParams.get("serverID");
+  }
+
+  async ping() {
+    await this.listContacts();
+  }
 
   newPerson(): ActiveSyncPerson {
     return new ActiveSyncPerson(this);
   }
-  newGroup(): ActiveSyncGroup {
-    return new ActiveSyncGroup(this);
+  newGroup(): never {
+    throw new NotSupported("ActiveSync does not support distribution lists");
+  }
+
+  async queuedSyncRequest(data: any): Promise<any> {
+    if (!this.syncState && !this.syncKeyBusy) try {
+      // First request must be an empty request.
+      this.syncKeyBusy = this.makeSyncRequest();
+      await this.syncKeyBusy;
+    } finally {
+      this.syncKeyBusy = null;
+    }
+    while (this.syncKeyBusy) try {
+      await this.syncKeyBusy;
+    } catch (ex) {
+      // If the function currently holding the sync key throws, we don't care.
+    }
+    try {
+      this.syncKeyBusy = this.makeSyncRequest(data);
+      return await this.syncKeyBusy;
+    } finally {
+      this.syncKeyBusy = null;
+    }
+  }
+
+  protected async makeSyncRequest(data?: any): Promise<any> {
+    let request = {
+      Collections: {
+        Collection: Object.assign({
+          SyncKey: this.syncState || "0",
+          CollectionId: this.serverID,
+        }, data),
+      },
+    };
+    let response = await this.account.callEAS("Sync", request);
+    if (!response) {
+      return null;
+    }
+    if (response.Collections.Collection.Status == "3") {
+      // Out of sync.
+      this.syncState = null;
+      await SQLAddressbook.save(this);
+    }
+    if (response.Collections.Collection.Status != "1") {
+      throw new EASError("Sync", response.Collections.Collection.Status);
+    }
+    this.syncState = response.Collections.Collection.SyncKey;
+    await SQLAddressbook.save(this);
```

Comment:
You're saving the sync state before we had a chance to parse and save the result. If this fails for some reason, we missed these changes or (worse) become incoherent with our local copy.

Instead, we need to:
* Lock (I guess that's what `queued()` does)
* Fetch
* Parse the result
* Set and save the sync key.
* Unlock

Reply:
Ugh, I've got the same pattern in `ActiveSyncAccount` for the global sync key, so I'll have to fix that first.

Reply:
Please give it some thought which cases can happen and how we should recover from them.

* If, for example, the processing of the changes throws reproducibly on a given change set, then if we don't save the sync key in case of error, the next attempt would run into the same error again and be stuck forever.
* If the error is temporarily, e.g. because the received data is corrupted on the wire, then we would want to re-try.

Reply:
Regarding the code, one way would be to wrap the async parsing function, so you could still keep saving the sync key in the same function here. Compare `IMAPFolder.runCommand()` and how it's called.

Reply:
I've filed #225 about this.